### PR TITLE
migrate current alpine importer to alpine importer-improver model

### PR DESCRIPTION
**Pull request by TG1999**
Signed-off-by: Tushar Goel <tushar.goel.dav@gmail.com>

### CODE REVIEW

**Comment by TG1999** [comment]
for issue #620

**Comment by Hritik14** [comment]
@TG1999 As the development in this branch is going on and we wanted to move ahead with #476, rebasing/merging this with/from main will likely cause merge conflicts. Please accept the incoming changes for the import statements.
Alternatively, give [kdiff3](http://kdiff3.sourceforge.net/) a try for solving merge conflicts, it's really smart.

**Comment by TG1999** [comment]
also solves issue #628 and #629

**Comment by TG1999** [comment]
@pombredanne @Hritik14 @sbs2001 Please check my current approach for the importer, if this looks good I will proceed with writing tests for this

File: vulnerabilities/improvers/default.py [hunk 1]
```diff
@@ -38,7 +39,7 @@ def get_inferences(self, advisory_data: AdvisoryData) -> Iterable[Inference]:
             )
 
 
-def get_exact_purls(affected_package: AffectedPackage) -> (List[PackageURL], PackageURL):
+def get_exact_purls(affected_package: AffectedPackage) -> Tuple[List[PackageURL], PackageURL]:
```
Comment (pombredanne) [inline vulnerabilities/improvers/default.py:42]: Move these changes in a separate PR.

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thank you. LGTM... just a few cosmetic nits for your consideration. Could you also add some tests?

File: vulnerabilities/importers/alpine_linux.py [hunk 2]
```diff
@@ -21,123 +21,182 @@
...
+
+            # TODO: Handle the CVE-????-????? case
+            qualifiers = {
+                "distroversion": distroversion,
+                "reponame": reponame,
+            }
+
+            aliases = []
+
+            if is_cve(vuln_ids[0]):
```
[the remark this answers is not in the document]
Comment (TG1999) [reply vulnerabilities/importers/alpine_linux.py:163]: We are splitting vulnerability ID and reference, so the first part of the split will give us ID, so checking for the first part of the split only.

File: vulnerabilities/importers/alpine_linux.py [hunk 3]
```diff
@@ -20,124 +20,192 @@
...
-                            Reference(
-                                reference_id=reference_id,
-                                url="https://support.zabbix.com/browse/{}".format(reference_id),
-                            )
+def fetch_response(url):
+    """
+    Fetch and return `response` from the `url`
+    """
+    response = requests.get(url)
+    if response.status_code == 200:
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:73]: Are there instances where the response is not 200, if so you'll like to document that (in a comment maybe) or we won't need this function.

File: vulnerabilities/importers/alpine_linux.py [hunk 4]
```diff
@@ -20,124 +20,192 @@
...
+        return response
+    raise Exception("Failed to fetch data from the URL")
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:75]: This exception is too broad

File: vulnerabilities/importers/alpine_linux.py [hunk 5]
```diff
@@ -20,124 +20,192 @@
...
+
+
+def fetch_advisory_directory_links(page_response_content):
+    """
+    Return `advisory_directory_links` present in `index_page`
+    """
+    index_page = BeautifulSoup(page_response_content, features="lxml")
+    alpine_versions = [link.text for link in index_page.find_all("a") if link.text.startswith("v")]
+
+    assert alpine_versions, f"No versions found in {BASE_URL}"
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:85]: Do we really need this explicit check ? Does this happen at all ?

File: vulnerabilities/importers/alpine_linux.py [hunk 6]
```diff
@@ -20,124 +20,192 @@
...
+    """
+    Fetch and return `response` from the `url`
+    """
+    response = requests.get(url)
+    if response.status_code == 200:
+        return response
+    raise Exception("Failed to fetch data from the URL")
+
+
+def fetch_advisory_directory_links(page_response_content):
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:78]:
```suggestion
def fetch_advisory_directory_links(page_response_content: str) -> List:
```

File: vulnerabilities/importers/alpine_linux.py [hunk 7]
```diff
@@ -20,124 +20,192 @@
...
+    """
+    Return `advisory_directory_links` present in `index_page`
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:80]:
```suggestion
    Return a list of advisory directory links present in ``page_response_content`` html string
```

File: vulnerabilities/importers/alpine_linux.py [hunk 8]
```diff
@@ -20,124 +20,192 @@
...
+    return advisory_directory_links
+
+
+def fetch_advisory_links(advisory_directory_page, advisory_directory_link):
+    """
+    Yield json file urls present in `advisory_directory_page`
+    """
+    advisory_directory_page = BeautifulSoup(advisory_directory_page, features="lxml")
+    anchor_tags = advisory_directory_page.find_all("a")
+    assert anchor_tags, f"No anchor tags found in {advisory_directory_link}"
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:98]: Maybe rather than failing entirely, we can have a logging message here ? Also, does this happen often ?

File: vulnerabilities/importers/alpine_linux.py [hunk 9]
```diff
@@ -20,124 +20,192 @@
...
+    for anchor_tag in anchor_tags:
+        if anchor_tag.text.endswith("json"):
+            yield f"{advisory_directory_link}{anchor_tag.text}"
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:101]: imo [urllib.parse.urljoin](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin) or `requests.compat.urljoin` will be a safer way to concat urls

File: vulnerabilities/importers/alpine_linux.py [hunk 10]
```diff
@@ -20,124 +20,192 @@
...
-        for p in record["packages"]:
-            advisories.extend(
-                self._load_advisories(
-                    p["pkg"],
-                )
-            )
-
+            record = fetch_response(link).json()
+            if not record["packages"]:
+                raise Exception(f'"packages" not found in {link}')
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:63]: Too broad exception. maybe a logger warning will suffice here

File: vulnerabilities/importers/alpine_linux.py [hunk 11]
```diff
@@ -20,124 +20,192 @@
...
+            yield f"{advisory_directory_link}{anchor_tag.text}"
+
+
+def process_record(record: dict) -> List[AdvisoryData]:
+    """
+    Return a list of AdvisoryData objects by processing data
+    present in that `record`
+    """
+    if not record["packages"]:
+        raise Exception(f'"packages" not found in this {record!r}')
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:110]: Too broad exception

File: vulnerabilities/importers/alpine_linux.py [hunk 12]
```diff
@@ -20,124 +20,192 @@
...
+
+
+def load_advisories(
+    pkg_infos: Mapping[str, Any],
+    distroversion: str,
+    reponame: str,
+    archs: List[str],
+) -> Iterable[AdvisoryData]:
+    """
+    Yields AdvisoryData by mapping data from `pkg_infos`
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:136]:
```suggestion
    Yield AdvisoryData by mapping data from `pkg_infos`
```

File: vulnerabilities/importers/alpine_linux.py [hunk 13]
```diff
@@ -20,124 +20,192 @@
...
+                "distroversion": distroversion,
+                "reponame": reponame,
+            }
+
+            affected_packages = []
+
+            try:
+                fixed_version = AlpineLinuxVersion(version)
+            except Exception as e:
+                raise Exception(f"{version!r} Not a valid Alpine Version") from e
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:179]: Too broad exceptions

File: vulnerabilities/importers/alpine_linux.py [hunk 14]
```diff
@@ -20,124 +20,192 @@
...
+            assert isinstance(archs, List)
```
Comment (Hritik14) [inline vulnerabilities/importers/alpine_linux.py:180]: This is an unnecessary assert

File: vulnerabilities/references.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    @classmethod
+    def from_id(cls, zbx_id):
+        """
+        Return a new ZbxReference from an ZBX-XXXX id.
+        """
+        if not zbx_id or not zbx_id.lower().startswith("zbx"):
+            return ValueError(f"Not a Zabbix advisory. Does not start with ZBX: {zbx_id!r}")
+        return cls(
+            reference_id=zbx_id,
+            url="https://support.zabbix.com/browse/{}".format(zbx_id),
```
Comment (Hritik14) [inline vulnerabilities/references.py:37]: use fstring

File: vulnerabilities/references.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    A Zabbix advisory reference. See https://support.zabbix.com
+    """
+
+    @classmethod
+    def from_id(cls, zbx_id):
+        """
+        Return a new ZbxReference from an ZBX-XXXX id.
+        """
+        if not zbx_id or not zbx_id.lower().startswith("zbx"):
+            return ValueError(f"Not a Zabbix advisory. Does not start with ZBX: {zbx_id!r}")
```
Comment (Hritik14) [inline vulnerabilities/references.py:34]:
```suggestion
            return ValueError(f"Not a Zabbix reference. Does not start with ZBX: {zbx_id!r}")
```

**Review by Hritik14 — CHANGES_REQUESTED** [review]
Thanks for the PR! I've marked a few things that I'd like you to consider.
Also, In general, there are a lot of asserts. IMO a loud logging mechanism should serve for those asserts than failing entirely and much of the asserts should make their place in the tests.

File: vulnerabilities/importers/alpine_linux.py (hunk 3 above)
Comment (TG1999) [reply vulnerabilities/importers/alpine_linux.py:73]: This is to ensure that we are able to fetch from the string that we are passing as URL, if we are not able to fetch anything from that we should know that and correct our code if needed

File: vulnerabilities/importers/alpine_linux.py [hunk 17]
```diff
@@ -20,124 +20,113 @@
...
 
-        return advisory_links
 
-    def updated_advisories(self) -> Set[Advisory]:
-        advisories = []
-        advisory_links = self.fetch_advisory_links()
-        for link in advisory_links:
-            advisories.extend(self._process_link(link))
+def fetch_advisory_links():
+    BASE_URL = "https://secdb.alpinelinux.org/"
```
[the remark this answers is not in the document]
Comment (TG1999) [reply vulnerabilities/importers/alpine_linux.py]: This is outdated, It is already on top

File: vulnerabilities/importers/alpine_linux.py (hunk 14 above)
Comment (TG1999) [reply vulnerabilities/importers/alpine_linux.py:180]: This is to ensure that the `archs` is a List only, and not any other dataType, imagine a case if archs is a str, and we would be iterating over every character of string rather than iterating over list of arch

File: vulnerabilities/improver.py [hunk 18]
```diff
@@ -27,10 +27,10 @@ class Inference:
     """
 
     vulnerability_id: str = None
-    aliases: List[str] = dataclasses.field(default_factory=list)
+    aliases: Optional[List[str]] = None
```
Comment (TG1999) [inline vulnerabilities/improver.py:30]:
```suggestion
    aliases: Optional[List[str]] = dataclasses.field(default_factory=list)
```

File: vulnerabilities/improver.py [hunk 19]
```diff
@@ -27,10 +27,10 @@ class Inference:
...
     confidence: int = MAX_CONFIDENCE
     summary: Optional[str] = None
-    affected_purls: List[PackageURL] = dataclasses.field(default_factory=list)
+    affected_purls: Optional[List[PackageURL]] = None
```
Comment (TG1999) [inline vulnerabilities/improver.py:33]:
```suggestion
    affected_purls: Optional[List[PackageURL]] = dataclasses.field(default_factory=list)
```
